corehq/util/couch.py

```diff
@@ -33,3 +36,130 @@ def categorize_bulk_save_errors(error):
+        self.error_ids = []
+
+    def __enter__(self):
+        self.to_save = []
+        self.to_delete = []
+        return self
+
+    def commit(self, cmd, docs):
+        try:
+            results = cmd(docs)
+        except BulkSaveError as e:
+            results = e.results
+            error_ids = {d['id'] for d in e.errors}
+            self.error_ids.extend(error_ids)
+            success_ids = [d['id'] for d in results
+                           if d['id'] not in error_ids]
+        else:
+            success_ids = [d['id'] for d in results]
+
+        if self.throttle_secs:
+            sleep(self.throttle_secs)
+        return success_ids
+
+    def commit_save(self):
+        success_ids = self.commit(self.db.bulk_save, self.to_save)
+        self.saved_ids.extend(success_ids)
+        self.to_save = []
+
+    def commit_delete(self):
+        success_ids = self.commit(self.db.bulk_delete, self.to_delete)
+        self.deleted_ids.extend(success_ids)
+        self.to_delete = []
+
+    def save(self, doc):
+        self.to_save.append(doc)
+        if len(self.to_save) >= self.chunksize:
+            self.commit_save()
+
+    def delete(self, doc):
+        self.to_delete.append(doc)
+        if len(self.to_delete) >= self.chunksize:
+            self.commit_delete()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.to_save:
+            self.commit_save()
+        if self.to_delete:
+            self.commit_delete()
+
+
+def iter_update(db, fn, ids):
+    """
+    Map `fn` over every doc in `db` matching `ids`
+
+    `fn` should accept a json dict from couch and return one of:
+     * a json dict (or wrapped couch document) - bulk save that document
+     * the string "DELETE" - bulk delete that document
+     * None - pass
+
+    iter_update returns an object with the following properties:
+    saved_ids, deleted_ids, ignored_ids,
+
+    Ex: mark dimagi users as cool, delete the Canadians, and ignore the rest
+
+        def mark_cool(user_dict):
+            user = CommCareUser.wrap(user_dict)
+            if user.is_dimagi:
+                user.is_cool = True
+                return user
+            if user.language == "Canadian":
+                return "DELETE"
+
+        iter_update(CommCareUser.get_db(), mark_cool, all_user_ids)
+
+    This looks up and saves docs in chunks and is intended for large changes
+    such as a migration.  If an id is not found, it is skipped.  In the
+    event of a BulkSaveError, it will re-process the unsaved documents.
+    Wrapping is optional, and this function will unwrap if needed.
+    """
+    IterResults = namedtuple('IterResults', ['ignored_ids', 'not_found_ids',
+                                             'deleted_ids', 'updated_ids'])
+    results = IterResults(set(), set(), set(), set())
+
+    def _iter_update(doc_ids):
+        with IterDB(db, chunksize=100) as iter_db:
+            for chunk in chunked(set(doc_ids), 100):
```

Comment:
Is this supposed to be `chunk_size` in the place of `100` `chunk in chunked(set(doc_ids), 100)`?